### PR TITLE
[FIRRTL] Implement SymbolUserOpInterface for InstanceOp

### PR DESCRIPTION
**Pull request by youngar**
This interface improves the performance of the instance op verifier by
using a pre-populated symbol table to look up the referenced module.

### CODE REVIEW

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 1]
```diff
@@ -1045,6 +1038,19 @@ static LogicalResult verifyInstanceOp(InstanceOp instance) {
...
+
+/// Verify the correctness of an InstanceOp.
+static LogicalResult verifyInstanceOp(InstanceOp instance) {
+
+  // Check that this instance is inside a module.
+  auto module = instance->getParentOfType<FModuleOp>();
+  if (!module) {
+    instance.emitOpError("should be embedded in a 'firrtl.module'");
+    return failure();
+  }
```
Comment (fabianschuiki) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:1052]: I stumbled across this some time ago and was wondering if this couldn't be moved into the `*.td` file as `HasParent` trait on the op?

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 2]
```diff
@@ -993,26 +993,19 @@ void InstanceOp::setAllPortAnnotations(ArrayRef<Attribute> annotations) {
...
-  if (!module) {
-    instance.emitOpError("should be embedded in a 'firrtl.module'");
-    return failure();
-  }
-
-  auto referencedModule = instance.getReferencedModule();
+LogicalResult InstanceOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
+  auto module = (*this)->getParentOfType<FModuleOp>();
+  auto referencedModule =
+      symbolTable.lookupNearestSymbolFrom<FModuleLike>(*this, moduleNameAttr());
```
Comment (fabianschuiki) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:999]: This is nice!

**Review by fabianschuiki — APPROVED** [review]
This looks very nice!

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp (hunk 1 above)
Comment (youngar) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:1052]: `HasParent` [only looks one level up](https://github.com/llvm/llvm-project/blob/387f95541bdc575f75bba95d323b51198d3b9e2b/mlir/include/mlir/IR/OpDefinition.h#L1215), while we need to be able to put these in `firrtl.when`s regions. Maybe we should add a new optrait?

**Comment by youngar** [comment]
Now we just need a version of this that can verify XMR usage :)

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp (hunk 1 above)
Comment (fabianschuiki) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:1052]: Ah good point!

**Comment by darthscsi** [comment]
Yay!

**Comment by lattner** [comment]
This is really really nice, can you do the same for the HW level?

**Comment by youngar** [comment]
Already done: https://github.com/llvm/circt/commit/e38fa006ba8063a2d11d3d9d2355bb073569e3db :)

**Comment by lattner** [comment]
awesome!

**Pull request merged**